### DISCUSSION TITLE
PyCall.jl py string equivalents

### DISCUSSION THREAD

**Discussion by icweaver** [cjdoris/PythonCall.jl#91] [Q&A] [answered]
Would it make sense to have a version of PyCall.jl's `py` string macros? The use case I see for this is wrapping your pre-written Python code and calling it from Julia, so maybe something like this:

![Screenshot from 2022-01-07 14-58-31](https://user-images.githubusercontent.com/25312320/148600090-e71cf221-d0ca-436e-9c0d-08077c4b62dd.png)

I know nothing about macros and namespace stuff, so could be totally off here!

**Comment by cjdoris** [comment, marked as answer]
We already have `@pyeval` and `@pyexec`: https://cjdoris.github.io/PythonCall.jl/stable/pythoncall/#PythonCall.pyeval

They're a little more verbose than PyCall's `@py_str` but also more general because you can extract values out of exec calls and convert the result types.

**Reply by icweaver** [reply]
Right, I think I was just using the functional versions of `@pyeval` and `@pyexec` in the example above. I could re-write the above as:

![Screenshot from 2022-01-07 17-27-15](https://user-images.githubusercontent.com/25312320/148615660-c406c479-ed63-492a-b581-c9a05aabdaf4.png)

but I guess I was wondering if their was an alternative way that you would recommend

**Reply by cjdoris** [reply]
What's wrong with this?
```
@pyexec """
import numpy as np
global py_sqrt
py_sqrt = np.sqrt
"""
py_sqrt(x) = @pyeval("py_sqrt")(x)
```

**Reply by icweaver** [reply]
Sorry for the toy example. So I think that's great when we just have smaller functions that we need to call. What I had more in mind was how do we handle something [like this](https://github.com/JuliaAstro/Transits.jl/blob/16fc5c91b01c5474da32a37fa6053258bca929a6/test/orbits/keplerian.jl#L290-L317)?

We have a bunch of Python functions that we would like to eventually add for our tests, and it just seems like needing to declare `global` for each one or wrapping them each with `pyexec`/`pyeval`s could get a bit cumbersome, but I could be misunderstanding the point here for how these should be used

**Reply by icweaver** [reply]
Actually, I think the "manual" version of the string macro idea is starting to grow on me:

```python
pyexec("""
import numpy as np
def py_sqrt(x):
	return np.sqrt(x)
def py_norm(x, y):
	return np.linalg.norm((x, y))
""", Main)
py_sqrt(x) = pyeval("py_sqrt", Main)(x)
py_norm(x, y) = pyeval("py_norm", Main)(x, y)
```

It's going to be a bit of a mental shift for me to get over `py""`, but I definitely do see the benefits of having the extending functionality that these alternatives bring. Applying conversions after the fact with something like:

```julia
py_sqrt(x) = pyeval(Float64, "py_sqrt($x)", Main)
```

feels especially slick. If this usage looks alright to you, then I'm totally happy to mark this as solved! I appreciate you taking the time =]

**Reply by cjdoris** [reply]
That usage looks fine.

Note that the python code is compiled each time you call `pyeval` whereas the macro `@pyeval` compiles it once for you. If performance is critical then you can `pycompile` the code and `pyeval` that instead.

**Reply by cjdoris** [reply]
Your main issue seems to be that PyCall's `@py_str` is evaluated in global scope whereas `@pyexec` uses a temporary local scope, so you don't clutter the global scope. I could easily add a global version of `@pyexec`.

**Reply by cjdoris** [reply]
Though personally I don't think it's a great hardship to just put a `global` statement at the top of `@pyexec`.

**Reply by icweaver** [reply]
I wasn't aware of that distinction between `pyeval` and `@pyeval`, thanks for pointing that out! I am struggling to implement this in the toy example, and think I am doing it in a really roundabout way:

```julia
@pyexec """
global py_sqrt
global np

import numpy as _np
np = _np

def _py_sqrt(x):
	return np.sqrt(x)
py_sqrt = _py_sqrt
"""
py_sqrt(x) = @pyeval("py_sqrt")(x)
```

> Though personally I don't think it's a great hardship to just put a global statement at the top of @pyexec.

haha, fair enough. I'm sure that this will feel like less of a hassle once I iron out my usage in the above example

**Reply by cjdoris** [reply]
I think this does what you want:
```python
global np, py_sqrt
import numpy as np
def py_sqrt(x):
    return np.sqrt(x)
```

**Reply by icweaver** [reply]
ugh, I guess that's what I get for only trying things out in a Pluto notebook. It turns out that `py_sqrt(x) = @pyeval("py_sqrt")(x)` needs to be in the same cell for `py_sqrt(3)` to work, otherwise it throws a NameError for the function, hence my over-complicated workaround above X(

That's my bad, thanks for your patience. I'm totally happy with the above and think having a global version of `@pyexec` would just be icing on the cake, but definitely not super critical!

**Reply by cjdoris** [reply]
Happy to help!
